chain/consensus/mir/conf_manager.go

```diff
@@ -31,75 +32,116 @@ var (
 	ReconfigurationVotesKey = datastore.NewKey("mir/reconfiguration-votes")
 )
 
+var _ client.Client = &ConfigurationManager{}
+
 type ConfigurationManager struct {
-	ctx context.Context // Parent context
-	ds  db.DB           // Persistent storage.
-	id  string          // Parent ID.
+	ctx           context.Context // Parent context
+	ds            db.DB           // Persistent storage.
+	id            string          // Parent ID.
+	nextReqNo     uint64          // Configuration number.
+	nextAppliedNo uint64          // Last applied number.
 }
 
-func NewConfigurationManager(ctx context.Context, ds db.DB, id string) *ConfigurationManager {
-	return &ConfigurationManager{
-		ctx: ctx,
-		ds:  ds,
-		id:  id,
+func NewConfigurationManager(ctx context.Context, ds db.DB, id string) (*ConfigurationManager, error) {
+	cm := &ConfigurationManager{
+		ctx:           ctx,
+		ds:            ds,
+		id:            id,
+		nextReqNo:     0,
+		nextAppliedNo: 0,
+	}
+	err := cm.recover()
+	if err != nil {
+		return nil, err
 	}
+	return cm, nil
 }
 
-// GetConfigurationData recovers configuration number, and configuration requests that may be not applied.
-func (c *ConfigurationManager) GetConfigurationData() (reqs []*mirproto.Request, nextReqNo uint64, err error) {
-	// o is the offset to distinguish cases when there are no applied configuration requests and when the
-	// applied reconfiguration request number is 0.
-	// Since we use uint64 we cannot return -1 when there are no applied configurations.
-	o := uint64(1)
+func (cm *ConfigurationManager) NewTX(_ uint64, data []byte) (*mirproto.Request, error) {
+	r := mirproto.Request{
+		ClientId: cm.id,
+		ReqNo:    cm.nextReqNo,
+		Type:     ConfigurationRequest,
+		Data:     data,
+	}
+
+	if err := cm.storeRequest(&r, cm.nextReqNo); err != nil {
+		log.With("validator", cm.id).Errorf("unable to store configuration request: %v", err)
+		return nil, err
+	}
+
+	// If a request with number n was stored then the stored configuration nonce can be no more than n+1.
+	// That is possible if a node crashes here.
+
+	cm.nextReqNo++
+	cm.storeNextConfigurationNumber(cm.nextReqNo)
 
-	nextReqNo = c.GetNextConfigurationNumber()
+	return &r, nil
+}
+func (cm *ConfigurationManager) Done(txNo t.ReqNo) error {
+	cm.nextAppliedNo = uint64(txNo) + 1
+	cm.storeAppliedConfigurationNumber(cm.nextAppliedNo)
+	cm.removeRequest(uint64(txNo))
+	return nil
+}
 
-	appliedNumber, found := c.GetAppliedConfigurationNumber()
-	if !found {
-		o = 0
+func (cm *ConfigurationManager) Pending() (reqs []*mirproto.Request, err error) {
+	for i := cm.nextAppliedNo; i < cm.nextReqNo; i++ {
+		r, err := cm.getRequest(i)
```

Comment:
`Pending` right now means that requests will be re-sent. And they will be resent in the same order as before.

Reply:
But if there are multiple pending, with holes in between. Like pending requests 1 and 3, but not 2. This seems to lead to an error. Can this happen? (I guess due to the FIFO mempool it can't...)

Reply:
No, that's not possible.